0026-an-extensible-db-design.adoc

```diff
@@ -0,0 +1,181 @@
+
+. Initial data of each swap when being created for a specific role for a specific alpha and beta to protocol where alpha and beta is one of: `HAN`, `HErc20`, `HALight`
+. Additional data added to the swap for a specific role (e.g. secret hash only known to Bob during announcement)
+. Additional data added to the swap's alpha or beta ledger data.
+
+While not necessarily in scope we considered the following as well:
+
+. Swap State (one per swap, aggregated state - record all states with timestamp)
+. Communication State (one per swap - record all states with timestamp)
+. Ledger State (two per swap, record all states with timestamp and additional information if applicable, e.g. transaction ID)
+
+
+== Final model
+
+Instead of starting from scratch, we start iterating over a slightly adapted table layout based on the one proposed by @bonomat
+https://3.basecamp.com/4403044/buckets/16118249/messages/2583518881#__recording_2600021433[in basecamp].
+
+The final model can be found at the end of this document.
+Our reasoning for creating certain table relations is summed up in the sections below.
+
+=== Swaps
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/14e942f9c5ea1791f01bb65db0e09ac11cf5ab4c/assets/db_schema/swap.puml&fmt=svg[Swaps]
+
+Our first challange was depicting `alpha` and `beta` protocols for the swap.
+Our first attempt was to not indicate this in the `swap` table but instead just use foreign key relationships.
+For that to work we would need to indicate in the child table whether it is `alpha` or `beta` (see below in `swap_protocol`).
+In order to reduce the amount of duplicated data we decided to add a separate table `protocol` and a `1 to many` relation ship from `swap` to `protocol`.
+`protocol` is a _static_ table.
+Static means, it contains all available protocols as identifiers.
+For now: `han_bitcoin`, `han_ethereum`, `herc20`, `halight`.
+
+This allows us to :
+
+* simple update the name if we decide to, i.e. instead of updating X rows in `swaps` we update 1 entry in `protocol`
+* add additional protocols
+
+The second thing which was a bit challenging was how to model data which is only known at a later point in time and how to model data which not all `Swaps` have in common.
+We decided to go for an _extension table_ approach which allows us to add additional information where needed:
+
+* Bob only learns the hash of the secret during the announcement, additionally not every Swap has a secret hash.
+* The shared swap id is only available after a swap was finalized.
+
+We added `secret_hash` and `shared_swap_id` as extension tables, that have the `swap_local_id` as primary key (so there can only be one entry per swap).
+This allows us to add information to one specific swap without having to update the `swap` table.
+Note that the actual secret is not stored in the database, Alice always computes it from the swap parameters and her seed.
```

Comment:
Is the secret hash ever stored for Alice?

Reply:
Good question. I don't know, I think she always derives it from the secret right?

Reply:
Yes, we always compute the secret for Alice (or extract it on Bob's side):

https://github.com/comit-network/comit-rs/blob/e6fc2daf2635cab3aff347045702f1e24d819c8d/cnd/src/swap_protocols/rfc003/alice/actions/generic_impl.rs#L77-L84

When loading `AcceptedSwap`s there is not secret (we just store `Request` and `Accept` messages atm):

https://github.com/comit-network/comit-rs/blob/7a5e2ab59fdf5e99ef545cbf780ab9254a8de876/cnd/src/load_swaps.rs#L8-L33

Reply:
She can definitely. My question was, did you intend to store it for Alice as-well?

Reply:
At the moment: no. 
My reasoning would be, that we have another source of security assumptions if we do that. At the moment there is only the seed and the local swap id are used to compute the actual secret. If we explicitly save the secret then there is a third "source" for where the secret is coming from. I'd avoid that.

Of course you could argue that, if Alice changes her seed (for whatever reason) she would lose the ability to compute historic swaps' secrets. But I wouldn't see a use-case where this is necessary...

Reply:
No one is talking about storing the secret?

But the answer to my question seems to be `no`, so I will consider this as resolved :)